src/rgw/rgw_sal_dbstore.cc

```diff
@@ -757,6 +757,566 @@ namespace rgw::sal {
+
+    int total_parts = 0;
+    int handled_parts = 0;
+    int max_parts = 1000;
+    int marker = 0;
+    uint64_t min_part_size = cct->_conf->rgw_multipart_min_part_size;
+    auto etags_iter = part_etags.begin();
+    rgw::sal::Attrs attrs = target_obj->get_attrs();
+
+    ofs = 0;
+    accounted_size = 0;
+    do {
+      ret = list_parts(dpp, cct, max_parts, marker, &marker, &truncated);
+      if (ret == -ENOENT) {
+        ret = -ERR_NO_SUCH_UPLOAD;
+      }
+      if (ret < 0)
+        return ret;
+
+      total_parts += parts.size();
+      if (!truncated && total_parts != (int)part_etags.size()) {
+        ldpp_dout(dpp, 0) << "NOTICE: total parts mismatch: have: " << total_parts
+		       << " expected: " << part_etags.size() << dendl;
+        ret = -ERR_INVALID_PART;
+        return ret;
+      }
+
+      for (auto obj_iter = parts.begin(); etags_iter != part_etags.end() && obj_iter != parts.end(); ++etags_iter, ++obj_iter, ++handled_parts) {
+        DBMultipartPart* part = dynamic_cast<rgw::sal::DBMultipartPart*>(obj_iter->second.get());
+        uint64_t part_size = part->get_size();
+        if (handled_parts < (int)part_etags.size() - 1 &&
+            part_size < min_part_size) {
+          ret = -ERR_TOO_SMALL;
+          return ret;
+        }
+
+        char petag[CEPH_CRYPTO_MD5_DIGESTSIZE];
+        if (etags_iter->first != (int)obj_iter->first) {
+          ldpp_dout(dpp, 0) << "NOTICE: parts num mismatch: next requested: "
+		  	 << etags_iter->first << " next uploaded: "
+			 << obj_iter->first << dendl;
+          ret = -ERR_INVALID_PART;
+          return ret;
+        }
+        string part_etag = rgw_string_unquote(etags_iter->second);
+        if (part_etag.compare(part->get_etag()) != 0) {
+          ldpp_dout(dpp, 0) << "NOTICE: etag mismatch: part: " << etags_iter->first
+			 << " etag: " << etags_iter->second << dendl;
+          ret = -ERR_INVALID_PART;
+          return ret;
+        }
+
+        hex_to_buf(part->get_etag().c_str(), petag,
+		  CEPH_CRYPTO_MD5_DIGESTSIZE);
+        hash.Update((const unsigned char *)petag, sizeof(petag));
+
+        RGWUploadPartInfo& obj_part = part->info;
+
+        ofs += obj_part.size;
+        accounted_size += obj_part.accounted_size;
+      }
+    } while (truncated);
+    hash.Final((unsigned char *)final_etag);
+
+    buf_to_hex((unsigned char *)final_etag, sizeof(final_etag), final_etag_str);
+    snprintf(&final_etag_str[CEPH_CRYPTO_MD5_DIGESTSIZE * 2],
+	     sizeof(final_etag_str) - CEPH_CRYPTO_MD5_DIGESTSIZE * 2,
+           "-%lld", (long long)part_etags.size());
+    etag = final_etag_str;
+    ldpp_dout(dpp, 10) << "calculated etag: " << etag << dendl;
+
+    etag_bl.append(etag);
+
+    attrs[RGW_ATTR_ETAG] = etag_bl;
+
+    if (compressed) {
+      // write compression attribute to full object
+      bufferlist tmp;
+      encode(cs_info, tmp);
+      attrs[RGW_ATTR_COMPRESSION] = tmp;
+    }
+
+    /* Update multipart object data with object name
```

Comment:
Can you update one of these comments so they're at least slightly different?  My belief is that one refers to the multipart head object and one refers to the multipart meta object, right?

Reply:
oops. yes..  I forgot to upload the updated comments.

Reply:
done.